### PR TITLE
NF: Adding a reset to default button for setting default directory path

### LINKED ISSUE #8013: [Bug] Not a valid path. Registration not possible.
###### Reproduction Steps

image

Everything works fine on my old cell phone. The above error occurs on the new Google Pixel 4a. Registration is not possible. When I started the app for the first time, I also allowed the phone to be shared.

Thanks for your help!

// Auf meinem alten Handy funktioniert alles. Auf dem neuen Google Pixel 4a kommt der o.g. Fehler. Eine Anmeldung ist nicht möglich. Beim erstmaligen Start der App habe ich auch die Freigabe auf den Handyspeicher zugelassen.

Danke für eure Hilfen!

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [x] character to confirm the points below:*

- [ ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by kartikeysaran**
## Purpose / Description
If a user enters incorrect Directory path then a dialog appears with an option to reset the path to default

## Fixes
Fixes #8013 

## How Has This Been Tested?
This has been tested on a Real Device 

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mrudultora** [comment]
```
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/03-dialogs.xml:123: Error: Duplicate string value "Reset to default", used in dialog_reset_to_default, reset_custom_buttons [DuplicateCrowdInStrings]
      <string name="dialog_reset_to_default">Reset to default</string>
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:178: Duplicates value in dialog_reset_to_default. Add a comment attribute to explain this duplication
```
Lint failure. Would you please check if there's already a string available for "Reset to default" ?

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 1]
```diff
@@ -327,7 +328,24 @@ private void initSubscreen(String action, PreferenceContext listener) {
...
+                                        dialog.dismiss();
+                                    }
+                                })
+                                .setNegativeButton(R.string.reset_custom_buttons, new DialogInterface.OnClickListener() {
+                                    public void onClick(DialogInterface dialog, int id) {
+                                        try {
+                                            CollectionHelper.initializeAnkiDroidDirectory("/sdcard/AnkiDroid");
+                                            collectionPathPreference.setText("/sdcard/AnkiDroid");
+                                        } catch (StorageAccessException storageAccessException) {
+                                            storageAccessException.printStackTrace();
```
Comment (mrudultora) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:344]: Use `Timber` here.

**Review by mrudultora — CHANGES_REQUESTED** [review]
Looks good to me. Small changes to be done.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 2]
```diff
@@ -327,7 +328,24 @@ private void initSubscreen(String action, PreferenceContext listener) {
                         return true;
                     } catch (StorageAccessException e) {
                         Timber.e(e, "Could not initialize directory: %s", newPath);
-                        Toast.makeText(getApplicationContext(), R.string.dialog_collection_path_not_dir, Toast.LENGTH_LONG).show();
+                        AlertDialog.Builder builder = new AlertDialog.Builder(Preferences.this);
+                        builder.setMessage(R.string.dialog_collection_path_not_dir)
+                                .setPositiveButton(R.string.dialog_ok, new DialogInterface.OnClickListener() {
+                                    public void onClick(DialogInterface dialog, int id) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:334]: nit: Can you use a lambda here

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 3]
```diff
@@ -327,7 +328,24 @@ private void initSubscreen(String action, PreferenceContext listener) {
...
+                                        dialog.dismiss();
+                                    }
+                                })
+                                .setNegativeButton(R.string.reset_custom_buttons, new DialogInterface.OnClickListener() {
+                                    public void onClick(DialogInterface dialog, int id) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:339]: nit: Can you use a lambda here

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 4]
```diff
@@ -327,7 +328,24 @@ private void initSubscreen(String action, PreferenceContext listener) {
...
+                                        try {
+                                            CollectionHelper.initializeAnkiDroidDirectory("/sdcard/AnkiDroid");
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:341]: use `CollectionHelper.getDefaultAnkiDroidDirectory`

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 5]
```diff
@@ -327,7 +328,24 @@ private void initSubscreen(String action, PreferenceContext listener) {
...
+                                            collectionPathPreference.setText("/sdcard/AnkiDroid");
+                                        } catch (StorageAccessException storageAccessException) {
+                                            storageAccessException.printStackTrace();
+                                        }
+                                    }
+                                });
+                        builder.create().show();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:348]: I /think/ you can just call `builder.show()`

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Looks good! Didn't think it'd be this easy

Could you check that you can minimise/restore the app and the functionality still works

**Comment by kartikeysaran** [comment]
> Could you check that you can minimise/restore the app and the functionality still works

Yes it works !

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java (hunk 2 above)
Comment (kartikeysaran) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:334]: Okay

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 6]
```diff
@@ -330,22 +330,12 @@ private void initSubscreen(String action, PreferenceContext listener) {
...
-                                        } catch (StorageAccessException storageAccessException) {
-                                            storageAccessException.printStackTrace();
-                                        }
-                                    }
-                                });
-                        builder.create().show();
+                                .setPositiveButton(R.string.dialog_ok, (dialog, which) -> dialog.dismiss())
+                                .setNegativeButton(R.string.reset_custom_buttons, (dialog, which) -> {
+                                        CollectionHelper.getDefaultAnkiDroidDirectory();
+                                        collectionPathPreference.setText("/sdcard/AnkiDroid");
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:336]:
```suggestion
                                        String defaultDirectory = CollectionHelper.getDefaultAnkiDroidDirectory();
                                        collectionPathPreference.setText(defaultDirectory);
```

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 7]
```diff
@@ -327,7 +328,14 @@ private void initSubscreen(String action, PreferenceContext listener) {
                         return true;
                     } catch (StorageAccessException e) {
                         Timber.e(e, "Could not initialize directory: %s", newPath);
-                        Toast.makeText(getApplicationContext(), R.string.dialog_collection_path_not_dir, Toast.LENGTH_LONG).show();
+                        AlertDialog.Builder builder = new AlertDialog.Builder(Preferences.this);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:331]: I think we'd be better with `MaterialDialog.Builder`

Comment (kartikeysaran) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:331]: We are using ``` setText() ``` that might not work with ``` MaterialDialog ```

**Review by david-allison-1 — APPROVED** [review]
Looks good to me!

**Review by mikehardy — APPROVED** [review]
Nice! even avoided a new string :sunglasses:

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for April 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in April)

Thanks!

**Pull request merged**